### PR TITLE
gall: fixes to %boon handling

### PR DESCRIPTION
**Pull request by belisarius222**
When receiving a `%lost` from Ames, Gall was sending a `%u` message to leave the subscription.  However, Gall wasn't calling the normal routine for this, so it wasn't adding the leave request to its request queue -- this was causing the request queue to become desynchronized, which at least partially contributed to the repeating "2nd watch ack" printfs.

This fix does not recover from a broken queue (at least not on my ship), but it should prevent a working queue from desynchronizing.  I'm not yet sure how we should try to recover from this condition, but it might be good to get this onto the network ASAP to prevent more ships from entering the error condition.

EDIT:
The first commit only fixed one of a few problems going on simultaneously.  https://github.com/urbit/urbit/pull/4454/commits/d317a0847b88e209b5c0e29dfb791aabf33b3889 fixed another one: Gall had been trying to %give a %done to Ames if it was unable to process a %boon from Ames.  This was the wrong thing to do, because Ames clients should not ack %boon's, and because Gall's attempt at acking actually delivered the %done back to itself, causing a later crash in wire parsing.  Now Gall crashes deliberately, Ames prints the stack trace, and then Gall (mostly) correctly enqueues a %leave to Ames in response.

This PR does not fix a serious underlying issue in Gall's handling of %boon's and %done's from Ames.  Gall does not properly handle acks to outgoing sequences of repeating %leave and %watch moves over Ames, leading to the "2nd-watch-ack" error message and potentially to more serious inconsistencies.

The solution to this problem is out of scope for this PR, but it will probably involve having Gall assign a nonce to each %watch message, which it will send to the publisher Gall, who will include the nonce in all %fact and %kick messages while that instance of the subscription is active.  When the client resubscribes, it will send a %watch with a new nonce.  This will allow the subscriber to ignore responses (%done's, %fact's, and %kick's) on an already-defunct nonce and prevent queue desynchronization.

### CODE REVIEW

**Comment by belisarius222** [comment]
cc @philipcmonk

**Review by joemfb — APPROVED** [review]
LGTM.

FWIW, I'm not sure whether it should be released before the rest of the "2nd watch ack" phenomenon is understood.

**Comment by belisarius222** [comment]
I agree.

—
~rovnys-ricfer
https://urbit.org

**Comment by philipcmonk** [comment]
Great catch!  When we receive a message with an empty `outstanding` queue, we go into the `%missing` branch here: https://github.com/urbit/urbit/blob/081c4579b268b8844feb6428a0ce28ccb8e0f516/pkg/arvo/sys/vane/gall.hoon#L663-L669

Basically, if we don't know what type of request it was, we acknowledge it both as a watch and a poke.  That could definitely cause "2nd watch ack".  You would see this in |verb because they'd both show up.

**Comment by belisarius222** [comment]
ready for re-review @joemfb

File: pkg/arvo/sys/vane/ames.hoon [hunk 1]
```diff
@@ -1909,6 +1909,9 @@
       =/  =bone  bone.shut-packet
       ::
       ?:  ?=(%& -.meat.shut-packet)
+        =+  ?~  dud  ~
+            %.  ~
+            (slog leaf+"ames: crashed on fragment" >mote.u.dud< tang.u.dud)
```
Comment (joemfb) [inline pkg/arvo/sys/vane/ames.hoon:1914]: I though we were going to print one line by default, and put the whole trace behind I flag -- do you think this happens rarely enough to print the full trace unconditionally?

**Review by joemfb — COMMENTED** [review]
The %gall changes all LGTM.

One question about %ames' output:

**Review by joemfb — APPROVED** [review]
LGTM

**Comment by philipcmonk** [comment]
Additional context on the `%done` misgiving and 2nd-watch-ack is here: #2851

**Pull request merged**